File: wikidata/concordances.php
<?php declare(strict_types=1);

include_once '../vendor/autoload.php';

$CC0 = new JSKOS\Concept([
    'uri' => 'http://creativecommons.org/publicdomain/zero/1.0/',
    'notation' => ['CC0']
]);
$VZG = new JSKOS\Concept([
    'prefLabel' => [
        'de' => 'Verbundzentrale des GBV',             
    ],
    'notation' => ['VZG'],
    'uri' => 'https://viaf.org/viaf/134648237'
]);

$concordances = [];
foreach (file('stats.csv', FILE_IGNORE_NEW_LINES) as $line) {
    list ($prop, $count) = explode(',', $line);

    $conc = new JSKOS\Concordance([
        'extent' => $count,
        'fromScheme' => [
            'uri' => 'http://bartoc.org/en/node/1940',
            'prefLabel' => [
                'en' => 'Wikidata'
            ]
        ],
        'publisher' => [ $VZG ],
        'license' => [ $CC0 ],
    ]);

    if (file_exists("$prop.csv")) {        
        $conc->modified = date('Y-m-d\TH:i:s',filemtime("$prop.csv"));
    }

    $mappings = [];
    foreach (['csv'=>'CSV', 'txt'=>'BEACON', 'ndjson'=>'JSKOS'] as $ext => $name) {        
        $file = "$prop.$ext";
        if (file_exists($file)) {
            $mappings[] = new JSKOS\Access([
                'download' => "https://coli-conc.gbv.de/concordances/wikidata/$file",
                'notation' => [$name]
            ]);
        }
    }
    $conc->mappings = $mappings;

    $concordances[$prop] = $conc;
}

foreach (file('properties.tsv', FILE_IGNORE_NEW_LINES) as $line) {
    $line = explode("\t", $line);
    $prop = preg_replace('/^.*(P[0-9]+)>$/','$1', $line[0]);
    if (isset($concordances[$prop])) {
        $concordances[$prop]->toScheme = [
            'uri' => "https://bartoc.org/en/node/".$line[2],
            'prefLabel' => [
                'en' => preg_replace('/^"|"@.+$/','',$line[3])
            ]
        ];
        $extent = preg_replace('/^"([0-9]+)".*/','$1',$line[5]);
        if ($extent) {
            $concordances[$prop]->toScheme->extent = $extent;
        }
        $concordances[$prop]->notation = [$prop];
        $concordances[$prop]->prefLabel = [
            'en' => preg_replace('/^"|"@.+$/','',$line[4])
        ];
/*        
        $concordances[$id]->WDPROPERTY = [
            'uri' => "http://www.wikidata.org/entity/$id",
            'prefLabel' => [ 
                'en' => preg_replace('/^"|"@.+$/','',$line[4])
            ],
            'notation' => [$id]
        ];
*/        
    }
}

usort($concordances, function($a, $b) {
    return $a->prefLabel['en'] <=> $b->prefLabel['en'];
});

$registry = new JSKOS\Registry([
    'prefLabel' => [
        'en' => 'Wikidata Mappings'
    ],
    'scopeNote' => [
        'en' => [
            'Mappings between Wikidata and other knowledge organization systems',
        ]
    ],
    'license' => [ $CC0 ],
    'publisher' => [ $VZG ],
    'concordances' => [
        [
            'set' => new JSKOS\Set($concordances)
        ]
    ]
]);

$file = 'wikidata-concordances.json';
file_put_contents($file, $registry->json());

File: wikidata/concordances.pl
#!/usr/bin/env perl
use v5.14;
use JSON::PP;
use POSIX qw(strftime);
use autodie;

my $CC0 = {
    uri      => 'http://creativecommons.org/publicdomain/zero/1.0/',
    notation => ['CC0']
};

my $VZG = {
    prefLabel => { de => 'Verbundzentrale des GBV' },
    notation  => ['VZG'],
    uri => 'https://viaf.org/viaf/134648237'
};

my %concordances;

open my $fh, 'stats.csv';
while ( my $line = <$fh> ) {
    chomp $line;
    my ( $prop, $count ) = split ',', $line;

    my %conc = (
        extend     => $count,
        fromScheme => {
            uri       => 'http://bartoc.org/en/node/1940',
            prefLabel => { en => 'Wikidata' },
            publisher => [$VZG],
            license   => [$CC0],
        }
    );

    if ( -r "$prop.csv" ) {
        $conc{modified} =
          strftime( '%Y-%m-%dT%H:%M:%S', gmtime( ( stat("$prop.csv") )[9] ) );
    }

    my @mappings;
    my %formats = ( csv => 'CSV', txt => 'BEACON', ndjson => 'JSKOS' );
    while ( my ( $ext, $name ) = each %formats ) {
        my $file = "$prop.$ext";
        if ( -e $file ) {
            push @mappings, {    # JSKOS access
                download =>
                  "https://coli-conc.gbv.de/concordances/wikidata/$file",
                notation => [$name]
            };
        }
    }

    $conc{mappings} = \@mappings;

    $concordances{$prop} = \%conc;
}

open my $fh, '<', 'properties.ndjson';
foreach (<$fh>) {
    my $row = decode_json($_);

    my $pid = $row->{prop}{value};
    unless ( $concordances{$pid} ) {
        say STDERR "unknown property $pid!";
        next;
    }

    $concordances{$pid}{notation} = [$pid];

    $concordances{$pid}{toScheme} = {
        uri       => "https://bartoc.org/en/node/" . $row->{bartoc},
        prefLabel => { en => $row->{kos}{label} }
    };

    # TODO: get sizeAt => toSchemeextent

    $concordances{$pid}{prefLabel} = { en => $row->{prop}{label} };
}

# usort($concordances, function($a, $b) {
#    return $a->prefLabel['en'] <=> $b->prefLabel['en'];
# });

my $registry = {
    prefLabel => { en => 'Wikidata Mappings' },
    scopeNote => {
        en => [
            'Mappings between Wikidata and other knowledge organization systems'
        ]
    },
    license      => [$CC0],
    publisher    => [$VZG],
    concordances => [ ( values %concordances ) ],
};

open my $fh, '>', 'wikidata-concordances.json';
say $fh JSON::PP->new->canonical->encode($registry);

